text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,595 @@
+
+### Span kind
+
+[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
+SHOULD be set according to the following table, based on the operation a span describes.
+
+| Operation name | Span kind|
+|----------------|-------------|
+| `publish`      | `PRODUCER`, if no `create` spans are present. `INTERNAL` otherwise. |
+| `create`       | `PRODUCER` |
+| `receive`      | `CONSUMER` |
+| `deliver`      | `CONSUMER` |
+| `settle`       | `INTERNAL` |
+
+Setting span kinds according to this table ensures, that span links between
+consumers and producers always go from a `PRODUCER` span on the producer side to
+a `CONSUMER` span on the consumer side. This allows analysis tools to interpret
+linked traces without the need for additional semantic hints.
+
+### Span relationships
+
+#### Producer spans
+
+"Publish" spans SHOULD be created for operations of providing messages for
+sending or publishing to an intermediary. A single "Publish" span can account
+for a single message, or for multiple messages (in the case of providing
+messages in batches). "Create" spans MAY be created. A single "Create" span
+SHOULD account only for a single message.
+
+If a "Create" span exists for a message, its context SHOULD be attached to
+the message. If no "Create" span exists, the context of the related "Publish"
+span SHOULD be attached to the message.
+
+#### Consumer spans
+
+##### Push-based scenarios
+
+"Deliver" spans SHOULD be created for operations of passing messages to the
+application when those operations are not initiated by the application
+code.
+
+##### Pull-based scenarios
+
+"Receive" spans SHOULD be created for operations of passing messages to the
+application when those operations are initiated by the application code.
+
+##### General considerations
+
+"Deliver" or "Receive" spans MUST NOT be created for messages which are not
+forwarded to the caller, but are pre-fetched or cached by messaging
+libraries or SDKs.
+
+A single "Deliver" or "Receive" span can account for a single message, for
+multiple messages (in case messages are passed for processing as batches), or
+for no message at all (if it is signalled that no messages were received).  For
+each message it accounts for, the "Deliver" or "Receive" span SHOULD link to
+the "Create" or "Publish" span for the message. In addition, if it is possible
+the "Create" or "Publish" span MAY be set as a parent of the "Deliver" or
+"Receive" span.
+
+#### Settlement spans
+
+"Settle" spans or events SHOULD be created for every manually or automatically
+triggered settlement operation. A single "Settle" span can account for a
+single message or for multiple messages (in case messages are passed for
+settling as batches). For each message it accounts for, the "Settle" span
+MAY link to the "Create" or "Publish" span for the message.
+
```

Comment:
wonder if it would be useful to mention that settle span represents different kinds of settlement: completion (ack), abandoning (nack), dead-lettering, etc

Reply:
I guess that would be interesting to users 🤔 . Would this be then an attribute on the settle span? I'm not that familiar with all but if the majority of "kinds of settlements" are common among messaging systems/sdks then we could add a enum for them. This could be a cool thing for back-ends to use to augment the UI experience, e.g., a nack was received.

Reply:
I suggest to handle those distinctions via attributes, it should be covered by https://github.com/open-telemetry/opentelemetry-specification/issues/3320.

Reply:
I linked the above-mentioned issue to this comment. Closing it here.